src/rust/engine/src/nodes.rs

```diff
@@ -868,13 +906,20 @@ impl WrappedNode for Task {
     let func = self.task.func;
     let entry = self.entry;
     let product = self.product;
+    let can_modify_workunit = self.task.can_modify_workunit;
 
     let result_val = externs::call(&externs::val_for(&func.0), &deps)?;
     match externs::get_type_for(&result_val) {
-      t if t == context.core.types.coroutine => {
-        Self::generate(context, params, entry, result_val).await
+      t if t == context.core.types.coroutine => Self::generate(context, params, entry, result_val)
+        .await
+        .map(|result_val| {
+          let maybe_new_level = Self::compute_new_workunit_level(can_modify_workunit, &result_val);
+          (result_val, maybe_new_level)
+        }),
+      t if t == product => {
+        let maybe_new_level = Self::compute_new_workunit_level(can_modify_workunit, &result_val);
+        Ok((result_val, maybe_new_level))
```

Comment:
It might still be good to break out the `t == product` and `compute_new_workunit_level` check into a second block, so that both generators and functions get the type checking? Not a blocker though. I think this was a bug before.